### PR TITLE
Cache provisioning status

### LINKED ISSUE #603: Cache Provisioning Status

### PR DESCRIPTION
**Pull request by pkudinov**
Checking provisioning status through Hasura takes 70-100ms on Dev. This PR caches the provisioning status inside of `Provisioner` and does make extra requests to Hasura on every run.

### CODE REVIEW

File: runner/src/provisioner/provisioner.ts [hunk 1]
```diff
@@ -44,6 +45,23 @@ export default class Provisioner {
       .replace(/\//g, '0');
   }
 
+  isUserApiProvisioned(accountId: string, functionName: string): boolean {
+    const accountIndexers = this.#hasBeenProvisioned[accountId];
+    if (!accountIndexers) { return false; }
+    return accountIndexers[functionName];
+  }
+
+  private setProvisioned(accountId: string, functionName: string) {
```
Comment (darunrs) [inline runner/src/provisioner/provisioner.ts:54]: Should return void. Also, I think Typescript complains if there isn't a space between the function name and the parenthesis. You'll want to do `isUserApiProvisioned (...)` and `setProvisioned (...)`.

File: runner/src/provisioner/provisioner.ts [hunk 2]
```diff
@@ -44,6 +45,23 @@ export default class Provisioner {
...
+    const accountIndexers = this.#hasBeenProvisioned[accountId] ?? {};
+    this.#hasBeenProvisioned = {
+      ...this.#hasBeenProvisioned,
+      [accountId]: {
+        ...accountIndexers,
+        [functionName]: true
+      }
+    }
```
Comment (darunrs) [inline runner/src/provisioner/provisioner.ts:62]: I think there's a missing semicolon here.

File: runner/src/indexer/indexer.ts [hunk 3]
```diff
@@ -86,7 +86,7 @@ export default class Indexer {
 
         if (options.provision && !indexerFunction.provisioned) {
           try {
-            if (!await this.deps.provisioner.isUserApiProvisioned(indexerFunction.account_id, indexerFunction.function_name)) {
+            if (!this.deps.provisioner.isUserApiProvisioned(indexerFunction.account_id, indexerFunction.function_name)) {
```
Comment (darunrs) [inline runner/src/indexer/indexer.ts:89]: This should be calling `fetchUserApiProvisioningStatus`, not `isUserApiProvisioned`. Or you can switch the names. Probably better to just switch the names or else you'll need to rename indexer tests.

File: runner/src/provisioner/provisioner.ts [hunk 4]
```diff
@@ -44,6 +45,23 @@ export default class Provisioner {
       .replace(/\//g, '0');
   }
 
+  isUserApiProvisioned(accountId: string, functionName: string): boolean {
```
Comment (darunrs) [inline runner/src/provisioner/provisioner.ts:48]: Considering this function is meant to be used in another function, this should become private.

**Review by darunrs — CHANGES_REQUESTED** [review]

File: runner/src/provisioner/provisioner.ts [hunk 5]
```diff
@@ -24,6 +24,7 @@ export interface DatabaseConnectionParameters {
 
 export default class Provisioner {
   tracer: Tracer = trace.getTracer('queryapi-runner-provisioner');
+  #hasBeenProvisioned: Record<string, Record<string, boolean>> = {};
```
[the remark this answers is not in the document]
Comment (pkudinov) [reply runner/src/provisioner/provisioner.ts:27]: I think this is a great answer: https://stackoverflow.com/a/59641565

TL;DR; – this makes the field truly private unlike the use of `private` keyword.

File: runner/src/provisioner/provisioner.ts (hunk 1 above)
Comment (pkudinov) [reply runner/src/provisioner/provisioner.ts:54]: Oh, my ESLint stopped working. I fixed this

File: runner/src/indexer/indexer.ts (hunk 3 above)
Comment (pkudinov) [reply runner/src/indexer/indexer.ts:89]: Correct, changed to `fetchUserApiProvisioningStatus`. Not sure why the tests should be changed?

File: runner/src/provisioner/provisioner.ts (hunk 4 above)
Comment (pkudinov) [reply runner/src/provisioner/provisioner.ts:48]: This is only made public for tests, which is not ideal of course, but does not require too many changes

File: runner/src/indexer/indexer.ts (hunk 3 above)
Comment (pkudinov) [reply runner/src/indexer/indexer.ts]: Got it, and fixed tests. I think `fetchUserApiProvisioningStatus` is better naming to reflect that this is an expensive call

File: runner/src/provisioner/provisioner.ts [hunk 6]
```diff
@@ -44,6 +45,23 @@ export default class Provisioner {
...
+
+  private setProvisioned (accountId: string, functionName: string): void {
+    const accountIndexers = this.#hasBeenProvisioned[accountId] ?? {};
+    this.#hasBeenProvisioned = {
+      ...this.#hasBeenProvisioned,
+      [accountId]: {
+        ...accountIndexers,
+        [functionName]: true
+      }
+    };
```
Comment (morgsmccauley) [inline runner/src/provisioner/provisioner.ts:62]:
```suggestion
    this.#hasBeenProvisioned[accountId] ??= {};
    this.#hasBeenProvisioned[accountId][functionName] = true;
```
We can use Nullish coalescing assignment (`??=`) to ensure the account record exists, and just mutate the single field rather than replacing the entire object.

**Review by darunrs — APPROVED** [review]
LGTM!
